docs/EVALUATION.md

```diff
@@ -0,0 +1,249 @@
+
+1) We want to run evaluations in an online fashion so we can use those evaluations for triggering policies. For example, we may want to trigger a policy that will retrain a model if the model's performance degrades below a certain threshold.
+
+2) We want to run evaluations in an offline fashion for experiments so we determine the performance of a model **[1] at a certain point in time** as well as the **[2] total pipeline performance** aggregated over the whole time of the pipeline run (= pipeline performance).
+
+## Setting
+
+We are looking at a completed pipeline run that produced multiple triggers and therefore multiple models. We want to evaluate the performance of each model.
+
+```mermaid
+gantt
+    title Pipeline run
+    dateFormat YYYY-MM-DD
+    axisFormat %Y
+    section models
+        Trigger1 :milestone, 1930, 0d
+        model 1          :a1, 1930, 10y
+        Trigger2 :milestone, :after a1
+        model 2          :a2, after a1, 5y
+        Trigger3 :milestone, :after a2
+        model 3          :a3, after a2, 5y
+        Trigger4 :milestone, :after a3
+        model 4          :a4, after a3, 25y
+        Trigger5 :milestone, :after a4
+        model 5          :a5, after a4, 25y
+        Trigger6 :milestone, :after a5
+        model 6          :a6, after a5, 5y
+        Trigger7 :milestone, :after a6
+        model 7          :a7, after a6, 5y
+```
+
+## Evaluation methodology
+
+An evaluation methodology consists of three main components:
+
+1) The evaluation metric (e.g. F1, accuracy, precision, recall, etc.)
+2) The evaluation data (set of samples out of the evaluation split used for single model evaluation)
+3) The points in time when the evaluations are run
+
+### Use case 1: online evaluation
+
+In this case we have to be extremely cautious to keep evaluation costs low as the are part of the trigger policy evaluation costs (or at least part of the core pipeline that cannot be disregarded like the experiment model evaluations).
+
+During the core pipeline execution (not including experiment mode pipeline steps), we cannot access future data for evaluation. Also we don't have an evaluation split available. Therefore, we need to use the training data up to the current point in time for evaluation. This is not ideal but it is the only way to evaluate the model in an online fashion.
+
+> The question of how far into the past we want to includes samples remains. More thought is needed before continuing on this.
+
+### Use case 2: offline evaluation
+
+In experiment mode we can potentially utilize the full evaluation split over the whole dataset time range. Being able to "look into the future" is important to determine the performance of a model.
+
+#### [2] Total pipeline performance
+
+We argue that the performance of pipeline w.r.t. a certain evaluation metric is the weighted average of the performance of the most up to date model at each point in time over the whole time range of the pipeline run.
```

Comment:
I don't think it's fully clear to me yet what exactly you want to average weightly. At each step in time, we have exactly _one_ latest model for which we can get `m` evaluation metrics (see my comment below). Do you want to aggregate those `m` values per model, do you want to aggregate all models (`n*m` values? if so, in what sense is this weighted)?

Reply:
In the whole document, I only looked at one arbitrary (but fixed) evaluation metric. When evaluating a pipeline with multiple metrics, we can just plot a different time series (and encode this 4th dimension with a color / or do a separate plot if it gets too messy). I'll integrate the idea of multiple evaluation metrics into the doc!

Reply:
Actually I was not talking about multiple metrics here. To clarify: With `m` metrics, I meant something like the matrix approach:   we evaluate each model m times (okay, then we can collect multiple metrics, but I was also thinking about a single metric like accuracy).

Reply:
so it was not clear to me -- assuming a single metric -- what to want to average. Great to include the multiple metric stuff, but my main concern is multiple metrics in the sense that we evaluate the 2010 model on 2006, 2007 etc, not that we get F1, Accuracy etc for the 2010 model